enable CAN communication

KantaTamura opened the pull request:
## 変更内容
- mcp2517fd用のデバイスツリーを追加
- CAN通信を行うためのカーネルモジュールの追加

## 変更詳細

### device-tree

作成した `mcp2517fd-overlay.dts` をカーネルのデバイスツリービルドフォルダに追加し，一緒にビルドするようにした

### kernel-module

MCP2517FDを用いたCAN通信を行うための最小限のカーネルモジュールをインストールするようにした
- spi_bcm2835 : raspberry pi zero 2 w で使用できるSPI通信用のモジュール
- can_dev : CAN Network Device Driver Interface
- can_raw : これが無いとCAN通信時に `address family not supported by protocol` と表示される
- mcp251xfd : MCP2517FD用のドライバ

## CAN通信を試す方法

### root login

'local.conf'に次の内容を加える

```
EXTRA_IMAGE_FEATURES:append = " debug-tweaks"
```

### install can-utils

`doc/zero/build.md` の内容に加えて，[meta-openembedded](https://github.com/openembedded/meta-openembedded) の `meta-oe` 追加し，'local.conf' に次の内容を加える．

```
IMAGE_INSTALL:append = " can-utils"
```

### candump & cangen

> ***note***
> can通信する相手がいないと `error-passive` となり，通信できない

rootログイン後下記のコマンドを試す
```
$ ip link set can0 up type can bitrate 1000000
$ candump can0 &
$ cangen can0
```

KantaTamura committed:
zero: add mcp2517fd devicetree

can controller (mcp2517fd) を認識するためのデバイス情報を追加
モジュールは未作成

Signed-off-by: KantaTamura <tkanta496@gmail.com>

sasataku commented:
PRの中にcommitが4つありますが、後半2つのcommitは、前半2つのcommitの修正なので、
git commit --amend を使って 前半2つのcommitに含めてしまうとよいです。
(詳細は会議で教えます)

sasataku commented on zero/meta-mcp2517fd/recipes-devicetree/linux/linux-raspberrypi/scsat1-rpi/mcp2517fd-overlay.dts:
```diff
@@ -0,0 +1,82 @@
+
+    /* enable spi for spi0.0 */
+	fragment@0 {
+			  target = <&spi0>;
```
indentが統一されていない
（全体的に)

KantaTamura replied:
notionでひな形を作成していたのでtabとspaceが入り乱れてますね...
tabの部分を修正しておきました！

sasataku commented on zero/meta-mcp2517fd/recipes-devicetree/linux/linux-raspberrypi/scsat1-rpi/mcp2517fd-overlay.dts:
```diff
@@ -0,0 +1,82 @@
+                compatible = "microchip,mcp2517fd";
+                pinctrl-names = "default";
+                pinctrl-0 = <&mcp2517fd_int_pin>;
+                spi-max-frequency = <20000000>;     // up to 20MHz SPI clock speed on mcp2517hd
```
typo.  
"mcp2517hd"

sasataku commented:
[質問]
レシピのディレクリ・ファイルの名前や、ネストの切り方とか、どのようなポリシーで決めてますか？



i.g)
recipes-devicetree/linux/linux-raspberrypi_%.bbappend
recipes-devicetree/linux/linux-raspberrypi/scsat1-rpi/mcp2517fd-overlay.dts

sasataku commented on zero/meta-mcp2517fd/conf/layer.conf:
```diff
@@ -0,0 +1,13 @@
+BBPATH .= ":${LAYERDIR}"
+
+# We have recipes-* directories, add to BBFILES
+BBFILES += "${LAYERDIR}/recipes-*/*/*.bb \
+            ${LAYERDIR}/recipes-*/*/*.bbappend"
```
meta-raspberrypiとかと同じ書き方にしていると思うけど、receipeが固定で決まっているなら、*を使わずに、個別に指定してもよいのでは？

sasataku commented on zero/meta-scsat1-rpi/conf/local.conf.sample:
```diff
@@ -14,9 +14,3 @@ BB_DISKMON_DIRS ??= "\
-RPI_EXTRA_CONFIG = "\
-dtparam=spi=on \n \
-dtoverlay=mcp2517fd \n \
-"
```
ここは、ひとつ前のcommit上で消したほうがいいかな。
（このcommitは、追加しているだけのcommitにする)

KantaTamura commented:
> [質問] レシピのディレクリ・ファイルの名前や、ネストの切り方とか、どのようなポリシーで決めてますか？
> 
> i.g) recipes-devicetree/linux/linux-raspberrypi_%.bbappend recipes-devicetree/linux/linux-raspberrypi/scsat1-rpi/mcp2517fd-overlay.dts

recipes-devicetreeディレクトリはrecipetool appendsrcfileを用いて生成したので，ディレクトリ名・構造はデフォルトの命名規則なのかな？と思っていました
```
recipetool appendsrcfile -wm scsat1-rpi ./meta-mcp2517fd virtual/kernel /tmp/mcp2517fd-overlay.dtso 'arch/${ARCH}/boot/dts/overlays/mcp2517fd-overlay.dts'
```
今回のMACHINE名はscsat1-rpiではなくraspberrypi0-2w-64なのでその部分は間違ってます...


meta-raspberrypiの[レシピ](https://github.com/agherzan/meta-raspberrypi/tree/master/recipes-core)を見ていたのですが，recipetool appendsrcfileでrecipe-devicetree下にlinuxというフォルダ単体が生成されているのは，レシピに複数の機能をもたせるためで，今回の場合は下記の様なファイル構造でも問題ないということであっていますか？(`layer.conf`は適切に修正)
```
meta-mcp2517fd
├─conf
│  ├─layer.conf
├─recipe-devicetree
   ├─file
   │  ├─mcp2517fd-overlay.dts
   ├─mcp2517fd-overlay.bbappend
```

`.bb`や`.bbappend`ファイルによく見れれる`_1.0.bb`や`_%.bb`となっているのはなぜでしょうか？

yashi commented:
> recipetool appendsrcfileでrecipe-devicetree下にlinuxというフォルダ単体が生成されているのは，レシピに複数の機能をもたせるためで，今回の場合は下記の様なファイル構造でも問題ないということであっていますか？

大丈夫。

多くの Layer が `bitbake-layers create-layer` で作ってんじゃないかな？「いらないだろ」っていうファイルが残っている repo も、たくさんあるよ。

昔作った layer だと最終的に↓こうなったけど、最初はフラットで追加するたびにディレクトリ作っていった (`proj` 部分だけ置換した)
```
$ tree -L 3 meta-proj
meta-proj
├── COPYING.MIT
├── README
├── conf
│   └── layer.conf
├── recipes-apps
│   ├── proj-app
│   │   └── proj-app_1.1.bb
│   └── proj-webapp
│       └── proj-webapp_1.0.bb
└── recipes-core
    ├── base-files
    │   └── base-files_%.bbappend
    ├── images
    │   └── proj-image.bb
    ├── systemd-conf
    │   ├── files
    │   └── systemd-conf_%.bbappend
    └── tzdata
        └── tzdata.bbappend

11 directories, 9 files
```

`%` は、 https://docs.yoctoproject.org/bitbake/singleindex.html#append-files

yashi commented on zero/meta-mcp2517fd/recipes-devicetree/linux/linux-raspberrypi_%.bbappend:
```diff
@@ -0,0 +1,6 @@
+SRC_URI += "file://mcp2517fd-overlay.dts;subdir=git/arch/${ARCH}/boot/dts/overlays"
```
自分なら patch にするかな。でも今回のはこのままで良いです。もし patch にしたかったら別の PRにしましょう。

yashi commented on zero/meta-mcp2517fd/conf/layer.conf:
```diff
@@ -0,0 +1,20 @@
+BBFILE_PATTERN_meta-mcp2517fd = "^${LAYERDIR}/"
+BBFILE_PRIORITY_meta-mcp2517fd = "6"
+
+LAYERDEPENDS_meta-mcp2517fd = "core"
```
`raspberrypi` に依存するのでは？
[`bitbake-layers`](https://docs.yoctoproject.org/next/dev-manual/layers.html#managing-layers) を使って、 layer をいじった後にカーネルだけビルドしてみると、どれに依存しているかわかるよ。

yashi commented on zero/meta-mcp2517fd/conf/layer.conf:
```diff
@@ -0,0 +1,20 @@
+BBFILES += "${LAYERDIR}/recipes-*/*/*.bb \
+            ${LAYERDIR}/recipes-*/*/*.bbappend"
+
+BBFILE_COLLECTIONS += "meta-mcp2517fd"
```
`meta-` は、外そうか。 ここから `meta-` はずしたら、下の変数の suffix も `meta-` を外さないとだめだと思う。 はずすのが一般的だと思ってたけど、 GitHub Search で探したらけっこう `layer` とか `meta`つけている人いた。
https://github.com/search?q=%2F%5EBBFILE_COLLECTIONS+%2F&type=code

yashi reviewed (changes requested):
修正が必要なのは依存関係だけかな。あとは趣味程度

KantaTamura committed:
zero: add kernel-module

mcp251xfd用のモジュールをインストールしている
spi_bcm2835 : spi通信
can-dev : can通信
mcp251xfd : mcp2517fd用モジュール
can-raw : 無いとエラー (address family not supported by protocol)

Signed-off-by: KantaTamura <tkanta496@gmail.com>

yashi commented:
LGTM
